[JARS-83] Support content region specification

hashknot opened the pull request:
Allow authorized users to specify content regions
through the resource page.

jdamerow commented on cookies/views/resource.py:
```diff
@@ -535,6 +545,64 @@ def edit_resource_details(request, resource_id):
+                'end_resource': _get_resource_by_id(request, content_region_end_resource),
+            })
+            content_region.save()
+            __part__ = Field.objects.get(uri='http://purl.org/dc/terms/isPartOf')
```
I would not use this relation, since it does not refer to another resource. Let's use extent here, see http://dublincore.org/documents/dcmi-terms/#terms-extent

jdamerow commented on cookies/views/resource.py:
```diff
@@ -535,6 +545,64 @@ def edit_resource_details(request, resource_id):
+            content_region_start_position = form.cleaned_data.get('content_region_start_position')
+            content_region_end_resource = form.cleaned_data.get('content_region_end_resource')
+            content_region_end_position = form.cleaned_data.get('content_region_end_position')
+            __document__ = Type.objects.get(uri='http://xmlns.com/foaf/0.1/Document')
```
where is this used?

jdamerow commented on cookies/views/resource.py:
```diff
@@ -535,6 +545,64 @@ def edit_resource_details(request, resource_id):
+                'target': resource,
+                'predicate': __part__,
+                'container': resource.container,
+                'sort_order': int(max_nr['sort_order__max'])+1,
```
this should probably be a sort order just for extent relationships

jdamerow commented on cookies/views/resource.py:
```diff
@@ -673,6 +741,51 @@ def resource_content(request, resource_id):
+    """
+
+    parent_resource = _get_resource_by_id(request, resource_id)
+    content_relations = parent_resource.content.filter(is_deleted=False, content_type='text/plain')
```
the content type should probably be part of the request (defaulting to text/plain), given that we probably want to support other formats in the future.

jdamerow commented on cookies/views/resource.py:
```diff
@@ -673,6 +741,51 @@ def resource_content(request, resource_id):
+    if not content_relations:
+        return HttpResponse('No text content found for resource \'%s\'.' % parent_resource.name)
+
+    content_relation = content_relations[0]
```
if there can be more than one then this should probably also be part of the request (defaulting to the first).

jdamerow commented on cookies/views/resource.py:
```diff
@@ -673,6 +741,51 @@ def resource_content(request, resource_id):
+        remote = get_remote(resource.external_source, resource.created_by)
+        target = resource.location
+        if resource.external_source == Resource.GILES:
+            target = _add_parameter(target, 'dw', 300)
```
the dw paramter should only be required when requesting images. Is this really necessary here?

jdamerow replied:
or better, this should only be required when content type is image

jdamerow commented on cookies/views/resource.py:
```diff
@@ -673,6 +741,51 @@ def resource_content(request, resource_id):
+        if not content:
+            content = remote.get(target)
+            if content:
+                cache.set(resource.location, content, None)
```
are responses from giles and the web not cached?

hashknot commented on cookies/forms.py:
```diff
@@ -231,6 +231,56 @@ class UserResourceURLForm(forms.Form):
+            'help_text': 'Specify content region\'s ending position',
+            'min_value': 0,
+        })
+        self.groups[1]['fields']['position'] = self['content_region_end_position']
```
Two reasons:
- The choices in resources list are based on the parent resource. If I define the resource field as a class variable, I won't be able to supply the choices specific to that resource.
- Because the text preview is highly dependent on start/end resource, position; I am establishing groups here and using the same in the template.

hashknot commented on cookies/models.py:
```diff
@@ -246,6 +246,14 @@ def state(self):
+    start_position = models.IntegerField()
+    start_resource = models.ForeignKey('Resource', related_name='resource_start')
+    end_position = models.IntegerField()
+    end_resource = models.ForeignKey('Resource', related_name='resource_end')
```
Not sure I understand. We don't want the content region to be deleted if the referenced start/end resource is deleted?

jdamerow replied:
in this case we probably do, but not the other way around. so I guess it depends in which direction the cascade is applied.

hashknot commented:
Addressed Julia's review comments:

- Added cancel button redirecting back to the resource page.
- Repurposed the existing content API for previewing text in content region definition form.
- Update content region form to directly fetch the data from remote sources.

jdamerow commented on cookies/templates/define_resource_content_region.html:
```diff
@@ -0,0 +1,127 @@
+{% block content %}
+
+<div class="row">
+    <div class="col-md-6">
```
Is there a reason you're not using the whole width of the page? Otherwise looks great!

hashknot replied:
All the form elements get a little too wide for reasonably higher resolutions. In all the instances I have seen, the preview text lines are never that long. I will still go ahead and change the class to `col-12 col-md-8` to use maximum width for lower resolutions and 8 columns (instead of 6) for higher resolutions.

Will that be okay?

diging-jenkins commented:
Sounds good
---
Julia Damerow, PhD
Scientific Software Engineer
ASU-SFI Center for Biosocial Complexity
Arizona State University
jdamerow@asu.edu





> On Sep 25, 2017, at 12:49 PM, Jitesh <notifications@github.com> wrote:
> 
> @hashknot commented on this pull request.
> 
> In cookies/templates/define_resource_content_region.html <https://github.com/diging/amphora/pull/80#discussion_r140879178>:
> 
> > +
> +{% load breadcrumbs %}
> +{% block breadcrumbs %}
> +<div class="breadcrumbs">
> +    <a href="{% url 'index' %}">{% trans 'Home' %}</a>
> +    {% breadcrumb_url 'Resources' 'resources' %}
> +    {% breadcrumb_url resource.name 'resource' resource.id  %}
> +    {% breadcrumb_url 'Content Regions' %}
> +    {% breadcrumb_url 'Define' 'define-resource-content-region' resource.id  %}
> +</div>
> +{% endblock %}
> +
> +{% block content %}
> +
> +<div class="row">
> +    <div class="col-md-6">
> All the form elements get a little too wide for reasonably higher resolutions. In all the instances I have seen, the preview text lines are never that long. I will still go ahead and change the class to col-12 col-md-8 to use maximum width for lower resolutions and 8 columns (instead of 6) for higher resolutions.
> 
> Will that be okay?
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub <https://github.com/diging/amphora/pull/80#discussion_r140879178>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AMg4wZyfjFhjGZD3rIek9UtH-wvOd_zFks5smAO3gaJpZM4PZot4>.
>

hashknot commented:
Addressed Julia's review comments:

- Parent resource is now listed as a choice while specifying content region's start/end position.
- Content region form uses maximum width for lower resolutions, and 8 columns for higher resolutions.

jdamerow commented:
These seem to be links but nothing happens when I click on them:
![image](https://user-images.githubusercontent.com/8881141/30930894-ec283d68-a377-11e7-9f5d-2d1653136135.png)

hashknot commented:
> These seem to be links but nothing happens when I click on them

Since we do not yet support extracting the content from a content region, they are links without hrefs. I thought I'll add the actual hrefs when that feature is added.
If you prefer not having dummy links, I can do this:

![screen shot 2017-09-27 at 12 06 29 pm](https://user-images.githubusercontent.com/1668396/30932449-5565ae74-a37c-11e7-80fb-64a8c94a214f.png) 

Thoughts?

jdamerow commented:
yes, that looks good to me.

hashknot commented:
The latest commit removes the dummy links.
